### PR TITLE
improve block propagation algo

### PR DESCRIPTION
**Pull request by damip**
There were inaccuracies, especially in the insert_wanted_block function

### CODE REVIEW

File: massa-protocol-worker/src/protocol_worker.rs [hunk 1]
```diff
@@ -226,27 +226,30 @@ mod nodeinfo {
...
+            while self.wanted_blocks.len() > max_node_wanted_blocks_size {
                 // remove oldest item
                 let (&h, _) = self
-                    .known_blocks
+                    .wanted_blocks
                     .iter()
                     .min_by_key(|(h, t)| (*t, *h))
                     .unwrap(); // never None because is the collection is empty, while loop isn't executed.
-                self.known_blocks.remove(&h);
+                self.wanted_blocks.remove(&h);
```
Comment (AureliaDolo) [inline massa-protocol-worker/src/protocol_worker.rs:246]: Maybe put that in a remove exceeding blocks method for clarity (and maybe not only here but also when ever this pattern is used)

**Review by AureliaDolo — APPROVED** [review]

File: massa-protocol-worker/src/protocol_worker.rs [hunk 2]
```diff
@@ -226,27 +226,30 @@ mod nodeinfo {
...
             }
+            self.insert_known_blocks(&[block_id], false, now, max_node_known_blocks_size);
         }
 
-        /// If given node previously manifested it wanted given block.
-        pub fn contains_wanted_block(&mut self, block_id: &BlockId) -> bool {
+        /// returns wheteher a node wants a block, and if so, updates the timestamp of that info to now()
```
Comment (gterzian) [inline massa-protocol-worker/src/protocol_worker.rs:251]: typo wheteher => whether

File: massa-protocol-worker/src/protocol_worker.rs [hunk 3]
```diff
@@ -226,27 +226,30 @@ mod nodeinfo {
...
                 let (&h, _) = self
-                    .known_blocks
+                    .wanted_blocks
                     .iter()
                     .min_by_key(|(h, t)| (*t, *h))
                     .unwrap(); // never None because is the collection is empty, while loop isn't executed.
-                self.known_blocks.remove(&h);
+                self.wanted_blocks.remove(&h);
             }
+            self.insert_known_blocks(&[block_id], false, now, max_node_known_blocks_size);
```
Comment (gterzian) [inline massa-protocol-worker/src/protocol_worker.rs:248]: A comment explaining why this is necessary would be helpful here.

**Review by gterzian — APPROVED** [review]
LGTM with nits

File: massa-protocol-worker/src/protocol_worker.rs (hunk 1 above)
Comment (damip) [reply massa-protocol-worker/src/protocol_worker.rs:246]: done

File: massa-protocol-worker/src/protocol_worker.rs (hunk 2 above)
Comment (damip) [reply massa-protocol-worker/src/protocol_worker.rs:251]: done

File: massa-protocol-worker/src/protocol_worker.rs (hunk 3 above)
Comment (damip) [reply massa-protocol-worker/src/protocol_worker.rs:248]: done

File: massa-protocol-worker/src/protocol_worker.rs (hunk 1 above)
Comment (AureliaDolo) [reply massa-protocol-worker/src/protocol_worker.rs:240]: I don't see the change you're talking about

**Review by gterzian — APPROVED** [review]

File: massa-protocol-worker/src/protocol_worker.rs (hunk 1 above)
Comment (damip) [reply massa-protocol-worker/src/protocol_worker.rs:240]: look at the absolute code. Since the code extracted by github for the preview didn't change and just moved to a new function, the change is not visible in the preview

**Comment by damip** [comment]
I corrected a typo

**Review by yvan-sraka — CHANGES_REQUESTED** [review]
What's the improvement proposed by this PR? I understand it's about fixing some correctness issue (and not a performance one). Do you have an edge case that would pass which this version of the algorithm and failed with the previous one? Could you turn it into a test?

**Comment by damip** [comment]
> What's the improvement proposed by this PR? I understand it's about fixing some correctness issue (and not a performance one). Do you have an edge case that would pass which this version of the algorithm and failed with the previous one? Could you turn it into a test?

Correctness is important ;)

Otherwise, 

> What's the improvement proposed by this PR? I understand it's about fixing some correctness issue (and not a performance one). Do you have an edge case that would pass which this version of the algorithm and failed with the previous one? Could you turn it into a test?

* correctness
* network optimization: not trying to ask for a block B from a node N if N has asked for this block recently (implying that they didn't have it)
* wanted_blocks was never cleared and only grew indefinitely until RAM exhaustion

**Comment by AureliaDolo** [comment]
> * wanted_blocks was never cleared and only grew indefinitely until RAM exhaustion

It was cleared when we receive a block, we sent it to the node that want it and remove that block from the wanted list.

File: massa-protocol-worker/src/protocol_worker.rs [hunk 4]
```diff
@@ -226,27 +226,42 @@ mod nodeinfo {
...
-        /// insert a block in wanted list of a node.
-        /// Note that it also insert the block as a not known block for that node.
-        pub fn insert_wanted_blocks(
-            &mut self,
-            block_id: BlockId,
-            max_node_wanted_blocks_size: usize,
-        ) {
-            self.wanted_blocks.insert(block_id, Instant::now());
-            while self.known_blocks.len() > max_node_wanted_blocks_size {
+        /// Remove the oldest items from known_blocks
```
Comment (AureliaDolo) [inline massa-protocol-worker/src/protocol_worker.rs:229]:
```suggestion
        /// Remove the oldest items from wanted_blocks
```

File: massa-protocol-worker/src/protocol_worker.rs [hunk 5]
```diff
@@ -226,27 +226,42 @@ mod nodeinfo {
...
+        ) {
+            // Insert into known_blocks
+            let now = Instant::now();
+            self.wanted_blocks.insert(block_id, now);
+            self.remove_excess_wanted_blocks(max_node_wanted_blocks_size);
+
+            // If the node wants a block, it means that it doesn't have it.
+            // To avoid asking the node for this block in the meantime,
+            // mark the node as not knowing the block.
+            self.insert_known_blocks(&[block_id], false, now, max_node_known_blocks_size);
```
Comment (AureliaDolo) [inline massa-protocol-worker/src/protocol_worker.rs:260]: You don't prune known blocks ?

**Review by AureliaDolo — APPROVED** [review]

File: massa-protocol-worker/src/protocol_worker.rs (hunk 5 above)
Comment (damip) [reply massa-protocol-worker/src/protocol_worker.rs:260]: the insert_known_blocks function does

Comment (damip) [reply massa-protocol-worker/src/protocol_worker.rs:267]: here I also separated remove_excess_known_blocks for more clarity

**Comment by damip** [comment]
> > * wanted_blocks was never cleared and only grew indefinitely until RAM exhaustion
> 
> It was cleared when we receive a block, we sent it to the node that want it and remove that block from the wanted list.
> 
> I think the assumption was that as everything goes wanted blocks is by design emptying itself, whereas known blocks are growing indefinitely
> 
> I think this is mostly useful to avoid adversarial nodes flooding the wanted blocks with random hashes.


I think there is a confusion between the blocks that we want, and the blocks that a node wants.

```
        /// The blocks the node asked for.
        wanted_blocks: Map<BlockId, Instant>,
```

This is out of our control, the node can ask for millions of random block IDs all the time and flood our RAM

**Review by AureliaDolo — DISMISSED** [review]

**Comment by yvan-sraka** [comment]
Could you write a test that discriminate previous and current behavior to prevent any regression?

**Review by adrien-zinger — DISMISSED** [review]

**Comment by damip** [comment]
> Could you write a test that discriminate previous and current behavior to prevent any regression?

added a regression test

**Review by AureliaDolo — APPROVED** [review]

**Review by yvan-sraka — APPROVED** [review]

**Comment by damip** [comment]
bors merge
